### DISCUSSION TITLE
Discussing a Command Line Interface

### DISCUSSION THREAD

**Discussion by sharky98** [AssetRipper/AssetRipper#1483] [Ideas]
Hi,

I have seen the issues #1343, #1296 and the pull #513. I understand the overhead of maintaining different interface is not something you want to do (hence not planned). And to be honest, the web UI works wonder when I was investigating a project.

However, now I would want to automate in some ways the import/export[^3] from another software. As a developer, it seems an overhead for me to boot up the web server and then make GET/POST API calls (that are not documented, yet?[^1]). A simpler way for me[^2], would be to either make a simple call to a CLI executable; or even have a DLL that expose the calls for the import/export.

I've looked at the code and I was able to figure out a way to create a very ultra basic CLI utility. It simply has two options, input (files/folders to import) and output (path to export), make sure the options are valid (path exist for input, path is empty for output) and redirect the process to [`AssetRipper.GUI.Web.GameFileLoader`](https://github.com/AssetRipper/AssetRipper/blob/1ae659e316ec30a369a1d91108596dbced729f8c/Source/AssetRipper.GUI.Web/GameFileLoader.cs) static class. It's basically
 - 40-ish lines for the core of the program;
 - 60-ish lines to create a compatible logger (could probably use the existing [`AssetRipper.Import.Logging.ConsoleLogger`](https://github.com/AssetRipper/AssetRipper/blob/1ae659e316ec30a369a1d91108596dbced729f8c/Source/AssetRipper.Import/Logging/ConsoleLogger.cs)).
 - 130-ish lines for the options and their validation (15-ish lines if we make no validation and leave it to `AssetRipper.GUI.Web.GameFileLoader`).

I could keep this on my side, my main issue is that `AssetRipper.GUI.Web.GameFileLoader` is not available as a library, so I must keep the CLI inside the solution[^4].

Thanks for reading me!

[^1]: With the code and real deep dive I was able to figure out how they works, but it was a pain.
[^2]: I know, not for you 😅
[^3]: In fact I would have even preferred to obtain the exported files directly from a GET or equivalent call to a DLL, instead of saving to disk and then reloading them to process what I want.
[^4]: Made me think, this could be an alternative where `AssetRipper.GUI.Web.GameFileLoader` is published as a library on nuget, so that us developers can use it in our own CLI or software?

**Comment by sharky98** [comment]
Here the source code of the basic CLI I made. I was careless on the naming of variable, files, classes and code style; so clearly not "production-ready". But it works 😂

<details>
<summary>Basic CLI code</summary>

## AssetRipper.CLI.csproj

```xml
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>net8.0</TargetFramework>
    <ImplicitUsings>enable</ImplicitUsings>
    <Nullable>enable</Nullable>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="Spectre.Console" Version="0.49.1" />
    <PackageReference Include="Spectre.Console.CLI" Version="0.49.1" />
  </ItemGroup>

  <ItemGroup>
    <ProjectReference Include="..\AssetRipper.Import\AssetRipper.Import.csproj" />
    <ProjectReference Include="..\AssetRipper.GUI.Web\AssetRipper.GUI.Web.csproj" />
  </ItemGroup>

</Project>
```

## Program.cs

```cs
using AssetRipper.CLI.command;
using AssetRipper.Import.Logging;
using Spectre.Console.Cli;
using AssetRipper.CLI;

CommandApp app = new CommandApp();

Logger.Add(new SpectreConsoleLogger());

app.Configure(config =>
{
	config.AddCommand<ImportExportCommand>("impexp");
});

return app.Run(args);
```

## SpectreConsoleLogger.cs

I suspect this one is probably not necessary and using `AssetRipper.Import.Logging.ConsoleLogger` should work, but I wanted to have some fun with the library Spectre 😉

```cs
using AssetRipper.Import.Logging;
using Spectre.Console;

namespace AssetRipper.CLI;

public class SpectreConsoleLogger : ILogger
{
	public void BlankLine(int numLines)
	{
		for (int i = 0; i < numLines; i++)
		{
			AnsiConsole.WriteLine();
		}
	}

	public void Log(LogType type, LogCategory category, string message)
	{
		Style mainStyle = new Style();

		switch (type)
		{
			case LogType.Info:
				mainStyle.Foreground(Color.Teal);
				break;

			case LogType.Debug:
				mainStyle.Foreground(Color.Navy);
				break;

			case LogType.Verbose:
				mainStyle.Foreground(Color.Grey);
				break;

			case LogType.Warning:
				mainStyle.Foreground(Color.Olive);
				mainStyle.Decoration(Decoration.Italic);
				break;

			case LogType.Error:
				mainStyle.Foreground(Color.Maroon);
				mainStyle.Decoration(Decoration.Bold);
				break;
		}

		Markup logMessage;
		if (category == LogCategory.None)
		{
			logMessage = Markup.FromInterpolated($"{message}", mainStyle);
		}
		else
		{
			logMessage = Markup.FromInterpolated($"[purple dim]{category}[/] : {message}", mainStyle);
		}

		AnsiConsole.Write(logMessage);
		AnsiConsole.WriteLine();
	}
}
```

## command/ImportExportCommand.cs

```cs
using Spectre.Console;
using Spectre.Console.Cli;
using AssetRipper.GUI.Web;

namespace AssetRipper.CLI.command;

public class ImportExportCommand : Command<ImportExportSettings>
{
	public override int Execute(CommandContext context, ImportExportSettings settings)
	{
		AnsiConsole.Write(new Rule("[red]Importing...[/]"));

		GameFileLoader.LoadAndProcess(settings.ImportPaths);

		AnsiConsole.Write(new Rule("[red]Exporting...[/]"));

		GameFileLoader.ExportPrimaryContent(settings.ExportPath);

		// Omitted
		return 0;
	}
}
```

## command/ImportExportSettings.cs

That's the big one where I simply validate the import path to make sure they simply exist and that the export path either does not exist or is empty (by prompting the user to confirm if not empty).

```cs
using Spectre.Console;
using Spectre.Console.Cli;

namespace AssetRipper.CLI.command;

public class ImportExportSettings : CommandSettings
{
	[CommandOption("-i|--importPath|--import <PATH>")]
	public required string[] ImportPaths { get; set; }

	[CommandOption("-e|--exportPath|--export")]
	public required string ExportPath { get; set; }

	public override ValidationResult Validate()
	{
		ValidationResult inputValidation = ValidateInput();
		if (!inputValidation.Successful)
		{
			return inputValidation;
		}

		ValidationResult outputValidation = ValidateOutput();
		if (!outputValidation.Successful)
		{
			return outputValidation;
		}

		return ValidationResult.Success();
	}

	private ValidationResult ValidateInput()
	{
		Rule rule = new("[red]Validate Import inputs[/]");
		rule.RuleStyle("red dim");
		AnsiConsole.Write(rule);
		AnsiConsole.WriteLine();

		bool hasAtLeastOnePath = ImportPaths.Length >= 1;
		bool doesPathsExist = ImportPaths.All(Path.Exists);

		if (!hasAtLeastOnePath)
		{
			return ValidationResult.Error("There must be at least one path.");
		}

		TextPath[] textPaths = GetTextPaths(ImportPaths);
		foreach (TextPath item in textPaths)
		{
			AnsiConsole.Write(item);
			AnsiConsole.WriteLine();
		}

		if (!doesPathsExist)
		{
			string[] wrongPaths = ImportPaths.Where(x => !Path.Exists(x)).ToArray();
			return ValidationResult.Error($"The following path(s) does not exist: {string.Join(", ", wrongPaths)}");
		}

		return ValidationResult.Success();
	}

	private ValidationResult ValidateOutput()
	{
		Rule rule = new("[red]Validate Export inputs[/]");
		rule.RuleStyle("red dim");
		AnsiConsole.Write(rule);
		AnsiConsole.WriteLine();

		bool doesPathsExist = Path.Exists(ExportPath);

		TextPath textPath = new TextPath(ExportPath)
				.RootColor(Color.Red)
				.StemColor(Color.Blue)
				.LeafColor(Color.Yellow);
		AnsiConsole.Write(textPath);
		AnsiConsole.WriteLine();

		if (!doesPathsExist)
		{
			AnsiConsole.MarkupLine("[bold deepskyblue3]The provided export path ({0}) does not exist.[/]", Markup.Escape(ExportPath));
			try
			{
				Directory.CreateDirectory(ExportPath);
				AnsiConsole.MarkupLine("[green3]{0} created.[/]", Markup.Escape(ExportPath));
			}
			catch (Exception ex)
			{
				return ValidationResult.Error($"Unable to create the directory: {ex.Message}");
			}

			return ValidationResult.Success();
		}

		AnsiConsole.MarkupLine("[darkgoldenrod]The provided export path ({0}) already exist.[/]", Markup.Escape(ExportPath));
		AnsiConsole.MarkupLine("[red3 bold]If you continue, its content will be deleted.[/]");
		bool confirm = AnsiConsole.Confirm($"Are you sure to delete the content of {Markup.Escape(ExportPath)}?", false);

		if (confirm)
		{
			try
			{
				Directory.Delete(ExportPath, true);
				Directory.CreateDirectory(ExportPath);
				AnsiConsole.MarkupLine("[green3]Content of {0} deleted.[/]", Markup.Escape(ExportPath));
				return ValidationResult.Success();
			}
			catch (Exception ex)
			{
				return ValidationResult.Error($"Unable to remove the content of the directory: {ex.Message}");
			}

		}
		return ValidationResult.Error($"User requested to not delete the content of {Markup.Escape(ExportPath)}.");
	}

	private TextPath[] GetTextPaths(IReadOnlyList<string> paths)
	{
		return paths.Select(x =>
		{
			TextPath textPath = new TextPath(x)
				.RootColor(Color.Red)
				.StemColor(Color.Blue)
				.LeafColor(Color.Yellow);

			return textPath;
		}).ToArray();
	}
}
```

</details>

**Reply by EnergyBun** [reply]
I'd like to use this, but I'm clueless as to what to do with this code.  What steps do I need to take to compile and/or use it and what are the arguments it accepts?

**Comment by ds5678** [comment]
Hello! This is a lot to unpack. I'm going to make separate messages.

**Comment by ds5678** [comment]
In regards to API documentation, I have thought about this. Part of #1446 might enable automated documentation of the API.

**Reply by ds5678** [reply]
@sharky98 f47a8694027e55e1bd13ebfd3341050a5fd21437

**Comment by ds5678** [comment]
In regards to obtaining files directly via requests, that is technically possible today. This functionality is used for asynchronous asset preview.

However, there's currently no convenient way for you to get a list of assets. I expect this to change when I implement asset paging and/or asset search.

**Comment by thesmartwon** [comment]
I would also like to automate running AssetRipper in a Linux VM as a build step for another tool. Sadly this is not easy because a web server's requests are a lot harder to use than a CLI. It also requires concurrently running a server and client which can be tricky.

Why was this project's simple CLI replaced with a complicated web server and GUI? I don't understand. If you want a low maintenance burden, why make a web server and GUI at all?

**Reply by ds5678** [reply]
AssetRipper always had a GUI.

**Comment by chloelcdev** [comment]
I would just like to pop in and say that a CLI would be nice for us over in my current home modding community as well. We would like to automate the process of getting a running unity project up for a specific game for modding purposes.

I understand making a separate CLI program could be a lot. Maybe we could just get some command line args? If we were just able to specify `assetripper.exe -o "game_path" "output_path"` that would be super helpful

**Comment by neoedmund** [comment]
Hi, I just made the CLI tool (using AI in 5 minutes, I didn't write any code, It just works) .

https://github.com/neoedmund/AssetRipperCLI
